libc-bottom-half/sources/preopens.c

```diff
@@ -219,13 +224,20 @@ int __wasilibc_find_abspath(const char *path,
     return fd;
 }
 
-/// This is referenced by weak reference from crt1.c and lives in the same
-/// source file as `__wasilibc_find_relpath` so that it's linked in when it's
-/// needed.
-// Concerning the 51 -- see the comment by the constructor priority in
-// libc-bottom-half/sources/environ.c.
-__attribute__((constructor(51)))
-static void __wasilibc_populate_preopens(void) {
+void __wasilibc_populate_preopens(void) {
+    // Fast path: If the preopens are already initialized, do nothing.
+    if (preopens_populated) {
+        return;
+    }
+
+    LOCK(lock);
```

Comment:
does this work with `_REENTRANT`?
this function calls internal_register_preopened_fd which tries to acquire the same lock.

Reply:
Indeed, good catch. I've now fixed that.